rust/src/ike/logger.rs

```diff
@@ -0,0 +1,287 @@
+        jb.set_uint("doi", doi as u64)?;
+    }
+    jb.set_bool("encrypted_payloads", tx.hdr.ikev1_header.encrypted_payloads)?;
+
+    if !tx.hdr.ikev1_header.encrypted_payloads {
+        // enable logging of collected state if not-encrypted payloads
+
+        // client data
+        jb.open_object("client")?;
+        if state.ikev1_container.client.key_exchange.len() > 0 {
+            jb.set_string(
+                "key_exchange_payload",
+                &state.ikev1_container.client.key_exchange,
+            )?;
+            if let Ok(client_key_length) =
+                u64::try_from(state.ikev1_container.client.key_exchange.len())
+            {
+                jb.set_uint("key_exchange_payload_length", client_key_length / 2)?;
+            }
+        }
+        if state.ikev1_container.client.nonce.len() > 0 {
+            jb.set_string("nonce_payload", &state.ikev1_container.client.nonce)?;
+            if let Ok(client_nonce_length) = u64::try_from(state.ikev1_container.client.nonce.len())
+            {
+                jb.set_uint("nonce_payload_length", client_nonce_length / 2)?;
+            }
+        }
+
+        jb.open_array("proposals")?;
+        for client_transform in &state.ikev1_container.client.transforms {
+            jb.start_object()?;
+            add_attributes(client_transform, jb)?;
+            jb.close()?;
+        }
+        jb.close()?; // proposals
+        jb.close()?; // client
+
+        // server data
+        jb.open_object("server")?;
+        if state.ikev1_container.server.key_exchange.len() > 0 {
+            jb.set_string(
+                "key_exchange_payload",
+                &state.ikev1_container.server.key_exchange,
+            )?;
+            if let Ok(server_key_length) =
+                u64::try_from(state.ikev1_container.server.key_exchange.len())
+            {
+                jb.set_uint("key_exchange_payload_length", server_key_length / 2)?;
+            }
+        }
+        if state.ikev1_container.server.nonce.len() > 0 {
+            jb.set_string("nonce_payload", &state.ikev1_container.server.nonce)?;
+            if let Ok(server_nonce_length) = u64::try_from(state.ikev1_container.server.nonce.len())
+            {
+                jb.set_uint("nonce_payload_length", server_nonce_length / 2)?;
+            }
+        }
+        jb.close()?; // server
+
+        jb.open_array("vendor_ids")?;
+        for vendor in state
+            .ikev1_container
+            .client
+            .vendor_ids
+            .union(&state.ikev1_container.server.vendor_ids)
+        {
+            jb.append_string(vendor)?;
+        }
+        jb.close()?; // vendor_ids
+    }
+    jb.close()?;
+
+    return Ok(());
+}
+
+fn append_payload_type_extended(js: &mut JsonBuilder, pt: &u8) -> Result<(), JsonError> {
+    match *pt {
+        1 => js.append_string(IsakmpPayloadType::SecurityAssociation.to_string().as_str())?,
```

Comment:
Is there a more idiomatic rust way to do this ?

Reply:
It looks like `IsakmpPayloadType` implements `FromPrimitive`, so it has a `from_u8` method that returns on option. That would make this method much smaller.

Reply:
Done. Much better:
```rust
fn append_payload_type_extended(js: &mut JsonBuilder, pt: &u8) -> Result<(), JsonError> {
    if let Some(v) = IsakmpPayloadType::from_u8(*pt) {
        js.append_string(&format!("{:?}", v))?;
    }
    Ok(())
}
```